File: V2rayNG/desktop/Lampa.Desktop/Services/SubscriptionUrlResolver.cs
using System.Text.RegularExpressions;

namespace Lampa.Desktop.Services;

/// <summary>
/// Rewrites managed /auto/{id} subscription URLs across the same fallback hosts as Android.
/// </summary>
public static class SubscriptionUrlResolver
{
    private static readonly Regex SubIdPath = new(@"/auto/([A-Za-z0-9_-]{6,128})/?", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    public static string? ExtractSubId(string? url)
    {
        var normalized = url?.Trim() ?? "";
        if (normalized.Length == 0) return null;
        var match = SubIdPath.Match(normalized);
        return match.Success ? match.Groups[1].Value : null;
    }

    public static bool IsManaged(string? url) => ExtractSubId(url) is not null;

    public static IReadOnlyList<string> CandidateUrls(string originalUrl)
    {
        var trimmed = originalUrl.Trim();
        var subId = ExtractSubId(trimmed);
        if (subId is null) return [trimmed];
        return new[]
        {
            AppEndpoints.SubscriptionPrimaryHost,
            AppEndpoints.SubscriptionFallbackHost,
            AppEndpoints.SubscriptionReserveHost,
        }.Select(host => BuildUrl(host, subId)).Distinct(StringComparer.OrdinalIgnoreCase).ToArray();
    }

    public static string NormalizeStoredUrl(string url)
    {
        var subId = ExtractSubId(url);
        return subId is null ? url.Trim() : BuildUrl(AppEndpoints.SubscriptionPrimaryHost, subId);
    }

    private static string BuildUrl(string host, string subId) => $"https://{host}/auto/{subId}";
}

File: V2rayNG/app/src/main/java/com/v2ray/ang/handler/SubscriptionUrlResolver.kt
package com.v2ray.ang.handler

import com.v2ray.ang.AppConfig
import java.net.URI

/** Resolves Хоттабыч subscription URLs with primary/fallback hosts. */
object SubscriptionUrlResolver {

    private val SUB_ID_PATH = Regex("/auto/([A-Za-z0-9_-]{6,128})/?")

    fun extractSubId(url: String?): String? {
        val normalized = url?.trim().orEmpty()
        if (normalized.isEmpty()) return null
        return SUB_ID_PATH.find(normalized)?.groupValues?.getOrNull(1)
    }

    fun isManagedSubscriptionUrl(url: String?): Boolean = extractSubId(url) != null

    /** Primary gw.zizmos.ru, then sub.subhotig.buzz, then v.hattabych.ru. */
    fun candidateUrls(originalUrl: String): List<String> {
        val trimmed = originalUrl.trim()
        val subId = extractSubId(trimmed) ?: return listOf(trimmed)
        return listOf(
            AppConfig.SUBSCRIPTION_PRIMARY_HOST,
            AppConfig.SUBSCRIPTION_FALLBACK_HOST,
            AppConfig.SUBSCRIPTION_RESERVE_HOST,
        ).map { buildUrl(it, subId) }.distinct()
    }

    fun normalizeStoredUrl(url: String): String {
        val subId = extractSubId(url) ?: return url.trim()
        return buildUrl(AppConfig.SUBSCRIPTION_PRIMARY_HOST, subId)
    }

    private fun buildUrl(host: String, subId: String): String = "https://$host/auto/$subId"
}
